src/ipa_pc/mod.rs

```diff
@@ -382,104 +382,125 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
+                (G::ScalarField::zero() - point),
+                G::ScalarField::one()
+            ]);
 
-        // Fresh random challenge x
-        fs_rng.absorb(&to_bytes![batch_commitment].unwrap());
-        let point: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
+            // z_i(X) = z(X) / (X - x_i)
+            let z_i_polynomial = &z_polynomial / &x_polynomial;
 
-        let mut computed_batch_v = G::ScalarField::zero();
+            // z_i(x)
+            let z_i_value = z_i_polynomial.evaluate(x_point);
 
-        for ((&v_i, y_i), x_i) in v_values.iter().zip(y_values).zip(points) {
+            // C_i' = C_i - y_i * G_0
+            let mut comm_dash = labeled_commitment.commitment().comm.clone();
+            comm_dash[0] = (comm_dash[0].into_projective() + &(-vk.comm_key[0].mul(y_i))).into_affine();
 
-            computed_batch_v = computed_batch_v + &(cur_challenge * &((v_i - &y_i) / &(point - x_i)));
+            // LC(C) = SUM ( lambda^i * z_i(x) * C_i' )
+            for (index, &comm_dash_segment) in comm_dash.iter().enumerate() {
+                if index == lc_comm.len() {
+                    lc_comm.push(G::zero());
+                }
+                lc_comm[index] = (lc_comm[index].into_projective()
+                    + &(comm_dash_segment.into_projective()
+                        .mul(&z_i_value)
+                        .mul(&cur_challenge)
+                    )
+                ).into_affine();
+            }
 
+            // lambda^i
             cur_challenge = cur_challenge * &lambda;
         }
 
-        let mut batch_values = vec![];
-        for labeled_commitment in commitments.iter() {
-            batch_values.push(batch_proof.batch_values[
-                *labels
-                    .get(labeled_commitment.label())
-                    .ok_or(Error::MissingEvaluation {
-                        label: labeled_commitment.label().to_string(),
-                    })?
-                ]
-            );
-        }
+        // z(x)
+        let z_x_value = z_polynomial.evaluate(x_point);
 
-        // Reconstructed v value added to the check
-        batch_values.push(computed_batch_v);
+        // LC(C) = SUM ( lambda^i * z_i(x) * C_i' ) - z(X) * C
+        lc_comm = lc_comm.iter().enumerate().map(|(index, &lc_comm_segment)| {
+            let h_comm_segment = if batch_proof.h_comm.len() > index {
+                batch_proof.h_comm[index]
+            } else {
+                G::zero()
+            };
+            (lc_comm_segment.into_projective() + &h_comm_segment.mul(-z_x_value)).into_affine()
+        }).collect::<Vec<_>>();
 
-        // The commitment to h(X) polynomial added to the check
-        let mut commitments = commitments;
+        // Expected open value added to the check
+        let batch_values = vec![G::ScalarField::zero()];
+
+        // The reconstructed commitment to h(X) polynomial put to the check
```

Comment:
// The reconstructed commitment to the  LC polynomial put to the check

Reply:
Done